src/auth/awsclient.ts

```diff
@@ -0,0 +1,261 @@
+// limitations under the License.
+
+import {GaxiosOptions} from 'gaxios';
+
+import {AwsRequestSigner} from './awsrequestsigner';
+import {
+  BaseExternalAccountClient,
+  BaseExternalAccountClientOptions,
+} from './baseexternalclient';
+import {RefreshOptions} from './oauth2client';
+
+/**
+ * AWS credentials JSON interface. This is used for AWS workloads.
+ */
+export interface AwsClientOptions extends BaseExternalAccountClientOptions {
+  credential_source: {
+    environment_id: string;
+    // Region can also be determine from the AWS_REGION environment variable.
+    region_url?: string;
+    // The url field is used to determine the AWS security credentials.
+    // This is optional since these credentials can be retrieved from the
+    // AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY and AWS_SESSION_TOKEN
+    // environment variables.
+    url?: string;
+    regional_cred_verification_url: string;
+  };
+}
+
+/**
+ * Interface defining the AWS security-credentials endpoint response.
+ */
+interface AwsSecurityCredentials {
+  Code: string;
+  LastUpdated: string;
+  Type: string;
+  AccessKeyId: string;
+  SecretAccessKey: string;
+  Token: string;
+  Expiration: string;
+}
+
+/**
+ * AWS external account client. This is used for AWS workloads, where
+ * AWS STS GetCallerIdentity serialized signed requests are exchanged for
+ * GCP access token.
+ */
+export class AwsClient extends BaseExternalAccountClient {
+  private readonly environmentId: string;
+  private readonly regionUrl?: string;
+  private readonly securityCredentialsUrl?: string;
+  private readonly regionalCredVerificationUrl: string;
+  private awsRequestSigner: AwsRequestSigner | null;
+  private region: string;
+
+  /**
+   * Instantiates an AwsClient instance using the provided JSON
+   * object loaded from an external account credentials file.
+   * An error is thrown if the credential is not a valid AWS credential.
+   * @param options The external account options object typically loaded
+   *   from the external account JSON credential file.
+   * @param additionalOptions Optional additional behavior customization
+   *   options. These currently customize expiration threshold time and
+   *   whether to retry on 401/403 API request errors.
+   */
+  constructor(options: AwsClientOptions, additionalOptions?: RefreshOptions) {
+    super(options, additionalOptions);
+    this.environmentId = options.credential_source.environment_id;
+    // This is only required if the AWS region is not available in the
+    // AWS_REGION environment variable
+    this.regionUrl = options.credential_source.region_url;
+    // This is only required if AWS security credentials are not available in
+    // environment variables.
+    this.securityCredentialsUrl = options.credential_source.url;
+    this.regionalCredVerificationUrl =
+      options.credential_source.regional_cred_verification_url;
+    const envIdComponents = this.environmentId?.match(/^(aws)([\d]+)$/) || [];
```

Comment:
`[\d]` can be replaced with `\d`. There is no need to have a character class here.

Reply:
Done